runtime/moonbase/src/migrations.rs

```diff
@@ -0,0 +1,75 @@
+use crate::{Balance, Runtime, AuthorMapping};
+
+/// The type responsible for migrating the moonbase runtime from ersion to version. Each migration
+/// is gated by runtime spec version, so only the correct migrations should be supplied. If you
+/// need to migrate storage, add the migration logic here.
+pub struct MoonbaseMigrationStation;
+
+impl OnRuntimeUpgrade for MoonbaseMigrationStation {
+	fn on_runtime_upgrade() -> Weight {
+
+		log::info!("🗺️ Running MoonbaseMigrationStation");
+
+        //TODO I'd like to detect the correct version that we ar emigrating _from_ here. This code
+        // seems to always give me the same version 43 from _this_ runtime though. Shawn's comments
+        // sound like that is expected.
+		// let previous_version: u32 = frame_system::LastRuntimeUpgrade::<Runtime>::get().map(|rv| rv.spec_version.into()).unwrap_or(0);
+        let previous_version = 40;
+
+		log::info!("🗺️ Previous version: {:?}", previous_version);
+		log::info!("🗺️ Upgrade version:  {:?}", crate::VERSION.spec_version);
+
+		match previous_version {
+			v if v < 40 => panic!("Migrating from runtimes below spec_version 40 is not supported!"),
+			v if v < 43 => migrate_author_mapping_storage_deposit_to_proper_units(),
+			_ => panic!("Cannot migrate from future runtime versions"),
+
+		}
+	}
+}
+
+fn migrate_author_mapping_storage_deposit_to_proper_units() -> Weight {
+	
+	// The previous version of the runtime had a security deposit of 100 wei.
+	// The new runtime uses 100 Units (with 18 decimals). To migrate, we iterate each
+	// registartion and reserve the additional required deposit. If an account does not have
+	// enough to cover the additional deposit, their registration is removed and their
+	// original deposit is refunded.
+
+	log::info!("🗺️ In helper");
+
+	let old_deposit: Balance = 100u32.into();
+	let additional_deposit = <Runtime as pallet_author_mapping::Config>::DepositAmount::get() - old_deposit;
+
+	// Iterate the entire storage mapping
+	pallet_author_mapping::Mapping::<Runtime>::translate(|_, account_id| {
+		// Try to reserve the additional amount
+		match <Runtime as pallet_author_mapping::Config>::DepositCurrency::reserve(&account_id, additional_deposit) {
+			// If they can afford it, return the same account_id to preserve the mapping
+			Ok(()) => {
+				log::info!("🗺️ Reserving more for account: {:?}", account_id);
+				Some(account_id)
+			},
+			// If they can't afford the additional deposit
+			Err(_) => {
+				log::info!("🗺️ Account can't afford deposit: {:?}", account_id);
+
+				// Refund the original deposit.
+				<Runtime as pallet_author_mapping::Config>::DepositCurrency::unreserve(&account_id, old_deposit);
+
+				// Return None to remove the registration from the map
+				None
+			}
+		}
+	});
+
+	log::info!("🗺️ about to return from helper");
+
+	//TODO No idea what weight I should be returning.
+	10_000u32.into()
```

Comment:
:eyes:

This makes me think the idea is to squeeze all migration into the first block upon an upgrade. That's quite a constraint, as this (and future migrations) might operate over extremely large and/or unpredictable sets of storage.

It's not even reasonable to make assumptions here based on current blockchain conditions. For example, we can't say "our upcoming migration should fit because there are only 200 items to migrate", because that may not be true in the future (and could be an attack vector anyway).

For reference, the current weight system assigns `25 usec` for reads and `75 usec` for writes: https://github.com/paritytech/substrate/blob/d8b3fce28513169d8ee1315dd9a4cd92d95753cc/frame/support/src/weights.rs#L164

Reply:
Yeah, that's a good point. I think the idea is to fit it into a single block and I agree that could have problems for migrating large data sets.

Substrate has a hook called on_idle that can be used to execute tasks in free block space. But we would have to suspend any operations on the data until it is migrated. That's not an option for the author mapping since we need it to check whether the author of the migration block is valid

I don't see your point about current blockchain conditions. "our upcoming migration should fit because there are only 200 items to migrate" is exactly the kind of logic I was hoping to use :) Why doesn't it work?